plugins/sigma/src/plugin.cpp

```diff
@@ -0,0 +1,211 @@
+#include <tenzir/arrow_table_slice.hpp>
+#include <tenzir/bitmap.hpp>
+#include <tenzir/concept/convertible/data.hpp>
+#include <tenzir/concept/convertible/to.hpp>
+#include <tenzir/concept/parseable/string/char_class.hpp>
+#include <tenzir/concept/parseable/tenzir/pipeline.hpp>
+#include <tenzir/concept/parseable/to.hpp>
+#include <tenzir/data.hpp>
+#include <tenzir/error.hpp>
+#include <tenzir/io/read.hpp>
+#include <tenzir/pipeline.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice_builder.hpp>
+
+#include <arrow/record_batch.h>
+#include <caf/error.hpp>
+#include <caf/expected.hpp>
+#include <caf/typed_event_based_actor.hpp>
+#include <fmt/format.h>
+
+#include <thread>
+
+namespace tenzir::plugins::sigma {
+
+class sigma_operator final : public crtp_operator<sigma_operator> {
+public:
+  sigma_operator() = default;
+
+  explicit sigma_operator(duration refresh_interval, std::string path)
+    : refresh_interval_{refresh_interval}, path_{std::move(path)} {
+  }
+
+  struct monitor_state {
+    [[nodiscard]] auto update(const std::filesystem::path& path)
+      -> caf::expected<void> {
+      if (std::filesystem::is_directory(path)) {
+        for (const auto& entry : std::filesystem::directory_iterator(path)) {
+          auto status = update(entry.path());
+          if (not status)
+            return status;
+        }
+        return {};
+      }
+      if (path.extension() != ".yml" && path.extension() != ".yaml") {
+        // We silently ignore non-yaml files.
+        return {};
+      }
+      auto query = tenzir::io::read(path);
+      if (not query) {
+        return std::move(query.error());
+      }
+      auto query_str = std::string_view{
+        reinterpret_cast<const char*>(query->data()),
+        reinterpret_cast<const char*>(query->data() + query->size())}; // NOLINT
+      auto yaml = from_yaml(query_str);
+      if (not yaml) {
+        return std::move(yaml.error());
+      }
+      auto rule = parse_rule(*yaml);
+      if (not rule) {
+        return std::move(rule.error());
+      }
+      rules[path.string()] = {std::move(*yaml), std::move(*rule)};
+      return {};
+    }
+
+    std::filesystem::path path;
+    std::unordered_map<std::string, std::pair<data, expression>> rules = {};
+  };
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    auto state = monitor_state{};
+    state.path = path_;
+    if (auto ok = state.update(state.path); not ok) {
+      diagnostic::error("failed to load sigma rules: {}", ok.error())
+        .emit(ctrl.diagnostics());
+    }
+    auto last_update = std::chrono::steady_clock::now();
+    co_yield {}; // signal that we're done initializing
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      if (last_update + refresh_interval_ < std::chrono::steady_clock::now()) {
+        if (auto ok = state.update(state.path); not ok) {
```

Comment:
Should successful file updates be logged?

Reply:
I can add a debug log for this, but I think everything else would be a bit too noisy.